### PR TITLE
Closes #73: A superuser should able to rerun automatic analysis via API

### LINKED ISSUE #73: A superuser should able to rerun automatic analysis via API

### PR DESCRIPTION
**Pull request by tomaszblazejczyk**

### CODE REVIEW

File: src/main/java/edu/harvard/iq/dataverse/api/MxrdrWorkflowApi.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+            AuthenticatedUser user = findAuthenticatedUserOrDie();
+            if (!user.isSuperuser()) {
+                return error(Response.Status.FORBIDDEN, "Only superusers can run workflow");
+            }
+            Dataset ds = findDatasetOrDie(id);
+            DatasetVersion updateVersion = null;
+            if (version != null) {
+                for (DatasetVersion datasetVersion:ds.getVersions()) {
+                    if (version.equals(datasetVersion.getVersionNumber() + "." + datasetVersion.getMinorVersionNumber())) {
+                        updateVersion = datasetVersion;
```
Comment (madryk) [inline src/main/java/edu/harvard/iq/dataverse/api/MxrdrWorkflowApi.java:96]: Are we sure that `updateVersion` couldn't be draft at this point?

File: src/main/java/edu/harvard/iq/dataverse/api/MxrdrWorkflowApi.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+                        rerun(dataset, released);
+                        datasetProcessed++;
+                    }
+                }
+            }
+            
+            return ok("Processed " + datasetProcessed + " datasets");
+        } catch (WrappedResponse ex) {
+            return ex.getResponse();
+        } catch (NoDatasetFilesException ex) {
```
Comment (madryk) [inline src/main/java/edu/harvard/iq/dataverse/api/MxrdrWorkflowApi.java:76]: In what situations that could happen?

File: src/main/java/edu/harvard/iq/dataverse/api/MxrdrWorkflowApi.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+            if (version != null) {
+                for (DatasetVersion datasetVersion:ds.getVersions()) {
+                    if (version.equals(datasetVersion.getVersionNumber() + "." + datasetVersion.getMinorVersionNumber())) {
+                        updateVersion = datasetVersion;
+                    }
+                }
+                if (updateVersion == null) {
+                    return error(Response.Status.BAD_REQUEST, "Unknown version: " + version);
+                } else {
+                    rerun(ds, updateVersion);
```
Comment (madryk) [inline src/main/java/edu/harvard/iq/dataverse/api/MxrdrWorkflowApi.java:102]: Correct me if I'm wrong but there is a possibility that we potentially end up with workflow in progress and dataset would not have any lock.
for example if we run this endpoint two times for the same dataset:
Run workflow nr 1:
1) add workflow lock
2) add execution to queue
Run workflow nr 2:
1) add workflow lock
2) add execution to queue
Processing:
1) workflow nr 1 finishes
2) workflow nr 1 removes locks
... dataset is unlocked ...
3) workflow nr 2 is executing

File: src/main/java/edu/harvard/iq/dataverse/api/MxrdrWorkflowApi.java [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+import edu.harvard.iq.dataverse.persistence.workflow.Workflow;
+import edu.harvard.iq.dataverse.persistence.workflow.WorkflowExecution;
+import edu.harvard.iq.dataverse.workflow.WorkflowServiceBean;
+import edu.harvard.iq.dataverse.workflow.execution.WorkflowContext;
+import edu.harvard.iq.dataverse.workflow.execution.WorkflowContext.TriggerType;
+import edu.harvard.iq.dataverse.workflow.execution.WorkflowExecutionFacade;
+import edu.harvard.iq.dataverse.workflow.execution.WorkflowExecutionService;
+
+@Path("mxrdr")
+public class MxrdrWorkflowApi extends AbstractApiBean {
```
Comment (madryk) [inline src/main/java/edu/harvard/iq/dataverse/api/MxrdrWorkflowApi.java:28]: I don't see the reason why this is not in the main dataverse repo. We don't use any `mxrdr` specific things here

File: src/main/java/edu/harvard/iq/dataverse/api/MxrdrWorkflowApi.java [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+                         "Unable to publish dataset, since there are no files in it.");
+        }
+
+    }
+    
+    private void rerun(Dataset dataset, DatasetVersion datasetVersion) throws WrappedResponse {
+        Optional<Workflow> workflow = workflowServiceBean.getDefaultWorkflow(PostPublishDataset);
+        if (workflow.isPresent()) {
+            workflowExecutionFacade.start(
+                    workflow.get(), new WorkflowContext(PostPublishDataset, dataset.getId(), datasetVersion.getMinorVersionNumber(), datasetVersion.getVersionNumber(), createDataverseRequest(findUserOrDie()), true));
```
Comment (madryk) [inline src/main/java/edu/harvard/iq/dataverse/api/MxrdrWorkflowApi.java:122]: Are we sure that rerun will not cause having merged dataset fields from two executions?

**Review by madryk — CHANGES_REQUESTED** [review]

File: src/main/java/edu/harvard/iq/dataverse/api/MxrdrWorkflowApi.java (hunk 4 above)
Comment (tomaszblazejczyk) [reply src/main/java/edu/harvard/iq/dataverse/api/MxrdrWorkflowApi.java:28]: Moved to https://github.com/CeON/dataverse/pull/1327

**Pull request closed without merging**